x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/vectors/TextEmbeddingQueryVectorBuilder.java

```diff
@@ -0,0 +1,138 @@
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+
+public class TextEmbeddingQueryVectorBuilder implements QueryVectorBuilder {
+
+    public static final String NAME = "text_embedding";
+
+    public static final ParseField MODEL_TEXT = new ParseField("model_text");
+
+    public static final ConstructingObjectParser<TextEmbeddingQueryVectorBuilder, Void> PARSER = new ConstructingObjectParser<>(
+        NAME,
+        args -> new TextEmbeddingQueryVectorBuilder((String) args[0], (String) args[1])
+    );
+
+    static {
+        PARSER.declareString(constructorArg(), InferModelAction.Request.MODEL_ID);
+        PARSER.declareString(constructorArg(), MODEL_TEXT);
+    }
+
+    public static TextEmbeddingQueryVectorBuilder fromXContent(XContentParser parser) throws IOException {
+        return PARSER.parse(parser, null);
+    }
+
+    private final String modelId;
+    private final String modelText;
+
+    public TextEmbeddingQueryVectorBuilder(String modelId, String modelText) {
+        this.modelId = modelId;
+        this.modelText = modelText;
+    }
+
+    public TextEmbeddingQueryVectorBuilder(StreamInput in) throws IOException {
+        this.modelId = in.readString();
+        this.modelText = in.readString();
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersion.V_8_7_0;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(modelId);
+        out.writeString(modelText);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(InferModelAction.Request.MODEL_ID.getPreferredName(), modelId);
+        builder.field(MODEL_TEXT.getPreferredName(), modelText);
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public void buildVector(Client client, ActionListener<float[]> listener) {
+        InferModelAction.Request inferRequest = InferModelAction.Request.forTextInput(
+            modelId,
+            TextEmbeddingConfigUpdate.EMPTY_INSTANCE,
+            List.of(modelText)
+        );
+
+        client.execute(InferModelAction.INSTANCE, inferRequest, ActionListener.wrap(response -> {
+            if (response.getInferenceResults().isEmpty()) {
+                listener.onFailure(new IllegalStateException("text embedding inference response contain no results"));
+                return;
+            }
+
+            if (response.getInferenceResults().get(0)instanceof TextEmbeddingResults textEmbeddingResults) {
+                listener.onResponse(textEmbeddingResults.getInferenceAsFloat());
+            } else {
```

Comment:
Could we accidentally return a `WarningInferenceResults`? It may be good to handle that case and extract the warning to throw an error.